src/Data/List/Kleene/Base.agda

```diff
@@ -0,0 +1,366 @@
+--   simplify certain proofs.
+
+infixr 5 _&_ ∹_
+mutual
+  -- Non-Empty Lists
+  record _⁺ {a} (A : Set a) : Set a where
+    inductive
+    constructor _&_
+    field
+      head : A
+      tail : A ⋆
+
+  -- Possibly Empty Lists
+  data _⋆ {a} (A : Set a) : Set a where
+    [] : A ⋆
+    ∹_ : A ⁺ → A ⋆
+open _⁺ public
+
+------------------------------------------------------------------------
+-- Uncons
+
+uncons : A ⋆ → Maybe (A ⁺)
+uncons []     = nothing
+uncons (∹ xs) = just xs
+
+------------------------------------------------------------------------
+-- FoldMap
+
+module _ {c ℓ} (sgrp : Semigroup c ℓ) where
+  open Semigroup sgrp
+
+  foldMap⁺ : (A → Carrier) → A ⁺ → Carrier
+  foldMap⁺ f (x & []) = f x
+  foldMap⁺ f (x & ∹ xs) = f x ∙ foldMap⁺ f xs
+
+module _ {c ℓ} (mon : Monoid c ℓ) where
+  open Monoid mon
+
+  foldMap⋆ : (A → Carrier) → A ⋆ → Carrier
+  foldMap⋆ f [] = ε
+  foldMap⋆ f (∹ xs) = foldMap⁺ semigroup f xs
+
+------------------------------------------------------------------------
+-- Folds
+
+module _ (f : A → B → B) (b : B) where
+  foldr⁺ : A ⁺ → B
+  foldr⋆ : A ⋆ → B
+
+  foldr⁺ (x & xs) = f x (foldr⋆ xs)
+
+  foldr⋆ [] = b
+  foldr⋆ (∹ xs) = foldr⁺ xs
+
+module _ (f : B → A → B) where
+  foldl⁺ : B → A ⁺ → B
+  foldl⋆ : B → A ⋆ → B
+
+  foldl⁺ b (x & xs) = foldl⋆ (f b x) xs
+
+  foldl⋆ b [] = b
+  foldl⋆ b (∹ xs) = foldl⁺ b xs
+
+------------------------------------------------------------------------
+-- Concatenation
+
+module Concat where
+  _⁺++⁺_ : A ⁺ → A ⁺ → A ⁺
+  _⁺++⋆_ : A ⁺ → A ⋆ → A ⁺
+  _⋆++⁺_ : A ⋆ → A ⁺ → A ⁺
+  _⋆++⋆_ : A ⋆ → A ⋆ → A ⋆
+
+  head (xs ⁺++⋆ ys) = head xs
+  tail (xs ⁺++⋆ ys) = tail xs ⋆++⋆ ys
+
+  xs ⋆++⋆ ys = foldr⋆ (λ x zs → ∹ x & zs) ys xs
+
+  xs ⁺++⁺ ys = foldr⁺ (λ x zs → x & ∹ zs) ys xs
+
+  []     ⋆++⁺ ys = ys
+  (∹ xs) ⋆++⁺ ys = xs ⁺++⁺ ys
+open Concat public using () renaming (_⁺++⁺_ to _++⁺_; _⋆++⋆_ to _++⋆_)
+
+------------------------------------------------------------------------
+-- Mapping
+
+module _  (f : A → B) where
+  map⁺ : A ⁺ → B ⁺
+  map⋆ : A ⋆ → B ⋆
+
+  head (map⁺ xs) = f (head xs)
+  tail (map⁺ xs) = map⋆ (tail xs)
+
+  map⋆ [] = []
+  map⋆ (∹ xs) = ∹ map⁺ xs
+
+module _ (f : A → Maybe B) where
+  mapMaybe⁺ : A ⁺ → B ⋆
+  mapMaybe⋆ : A ⋆ → B ⋆
+
+  mapMaybe⁺ (x & xs) with f x
+  mapMaybe⁺ (x & xs) | just y  = ∹ y & mapMaybe⋆ xs
+  mapMaybe⁺ (x & xs) | nothing = mapMaybe⋆ xs
+
+  mapMaybe⋆ [] = []
+  mapMaybe⋆ (∹ xs) = mapMaybe⁺ xs
+
+------------------------------------------------------------------------
+-- Applicative Operations
+
+pure⁺ : A → A ⁺
+head (pure⁺ x) = x
+tail (pure⁺ x) = []
+
+pure⋆ : A → A ⋆
+pure⋆ x = ∹ pure⁺ x
+
+module Apply where
+  _⋆<*>⋆_ : (A → B) ⋆ → A ⋆ → B ⋆
+  _⁺<*>⋆_ : (A → B) ⁺ → A ⋆ → B ⋆
+  _⋆<*>⁺_ : (A → B) ⋆ → A ⁺ → B ⋆
+  _⁺<*>⁺_ : (A → B) ⁺ → A ⁺ → B ⁺
+
+  []     ⋆<*>⋆ xs = []
+  (∹ fs) ⋆<*>⋆ xs = fs ⁺<*>⋆ xs
+
+  fs ⁺<*>⋆ xs = map⋆ (head fs) xs Concat.⋆++⋆ (tail fs ⋆<*>⋆ xs)
```

Comment:
The modules where the binary operators are defined do 4 separate functions (`*op+`, `*op*`, `+op*`, and `+op+`) but I thought it would be kind of confusing to export all of them, so we instead just export `*op*` and `+op+`, renamed to `op*` and `op+`. 

Here, this can change to `++⋆`, but below we have to use the qualified version.